streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -204,7 +203,14 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
             } else if (standbyTasks.containsKey(task.id()) && !task.isActive()) {
                 task.resume();
                 standbyTasksToCreate.remove(task.id());
-            } else /* we previously owned this task, and we don't have it anymore, or it has changed active/standby state */ {
+            } else {
+                // check for tasks that were owned previously but have changed active/standby status
+                final boolean isTransitioningType = activeTasks.containsKey(task.id()) || standbyTasks.containsKey(task.id());
+                if (isTransitioningType) {
```

Comment:
This is a meta comment: I'm a bit inclined to suggest that we do this recycle on the task-manager, i.e. do not call task.close() inside task-manager and let the close() internally check a boolean whether it is prepareRecycle. 

Instead, we can do the following:

1. Add a `Task#convertTo(TaskType)` interface which would return an active / standby task copying the fields of the original task, originated in `RESTORING` state.

2. For active task, the implementation would be:

* first transit to the `RESTORING` state (we would allow `SUSPENDED` to transit to `RESTORING` too, so if it is not in CREATED, we can first suspend it and then transit to `RESTORING`).

* and then return a newly created standby task initialized state as `RESTORING`.

3. For standby task, the implementation would be:

* first transit to `RESTORING` state (which would usually be a no-op).

* and then return a newly created active task initialized state as `RESTORING`.

Also I realized that `recordCollector.initialize();` in active task should be moved from `initializeIfNeeded` to `completeRestoration`. This is a minor bug that may block this proposal --- I will prepare a PR fixing this.

4. Then on task manager, for those convertible tasks we would call `convertTo` instead of close / re-create via the task-creators.

The key behind this proposal is that:

* Suspended and Restoring states are actually the same for active and standby tasks. 
* In the future when we remove `Suspended` state we would just have Restoring.
* Active and Standby's Restoring state are actually the same in terms of functionality.

Reply:
Yeah this does sound appealing, and is more or else what I first tried out when I started work on this (a long long time ago). My take at the time was that it added some complexity to the already-complicated TaskManager#handleAssignment while gaining little benefit, and was less "future-proof" compared to the current approach. For example it seems safer to just reuse the bare minimum of the task needed to recycle the state rather than rely on assumptions about which task states might be functionally equivalent to which states of the other type. It also seemed weird to require each task know how and be able to create a task of the opposite type, I felt this was more appropriate for the task creators.